signbank/dictionary/adminviews.py

```diff
@@ -249,6 +252,69 @@ def subquery_render_to_csv_response(self, context):
 
         return response
 
+    def ready_for_validation_render_to_csv_response(self, context):
+        if not self.request.user.has_perm("dictionary.export_csv"):
+            msg = _("You do not have permissions to export to CSV.")
+            messages.error(self.request, msg)
+            raise PermissionDenied(msg)
+
+        # Create the HttpResponse object with the appropriate CSV header.
+        # It is a Python file-like object.
+        response = HttpResponse(content_type="text/csv; charset=utf-8")
+        response[
+            "Content-Disposition"
+        ] = 'attachment; filename="ready-for-validation-export.csv"'
+
+        # Set up for outputting CSV.
+        writer = csv.writer(response)
+
+        # The queryset may or may not already be filtered for the ready for validation tag.
+        # We have to make sure it is filtered by the tag, so we are filtering again
+        ready_for_validation_qs = TaggedItem.objects.get_union_by_model(
+            self.get_queryset(),
+            Tag.objects.filter(name=settings.TAG_READY_FOR_VALIDATION)
+        )
+
+        csv_queryset = (
+            ready_for_validation_qs
+            .prefetch_related("glossvideo_set", "glosstranslations_set")
+            .annotate(
+                gloss_main_aggregate=StringAgg(
+                    GlossTranslations.objects.filter(
+                        gloss=OuterRef("pk"),
+                        language__language_code_2char="EN"
+                    ).values("translations")[:1],
+                    self.CSV_AGG_DELIM,
+                    distinct=True,
+                    output_field=CharField()
+                ),
+            )
+        )
+
+        headers = [
+            "idgloss",
+            "gloss_main",
+            "video_url"
+        ]
+        writer.writerow(headers)
+
+        for gloss_record in csv_queryset:
+            row = [gloss_record.idgloss, gloss_record.gloss_main_aggregate]
+            validation_video = gloss_record.glossvideo_set.filter(
```

Comment:
**suggestion**: add this lookup to the `prefetch_related` call above

As-is, this code will not use the prefetch because it adds a new filter. However, prefetch_related supports filtering using the [Prefetch](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#django.db.models.Prefetch) object

If you create a queryset for `GlossVideo` which has this filter, it can be used to narrow down which rows will be prefetched.

You will then be able to iterate over `glossvideo_set` and it will use the prefetch with the filter, resulting in 1 query instead of n

Reply:
TIL 🙏🏻  I never actually had a usecase for `Prefetch` before